File: faunadb-scala/src/main/scala/faunadb/streaming/SnapshotEventFlowProcessor.scala
package faunadb.streaming

import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.{Flow, SubmissionPublisher}

import faunadb.values._
import org.slf4j.LoggerFactory

import scala.concurrent.{ExecutionContext, Future}
import scala.util.{Failure, Success}

private [faunadb] class SnapshotEventFlowProcessor(loadDocument: () => Future[Value])(implicit ec: ExecutionContext)
  extends SubmissionPublisher[Value] with Flow.Processor[Value, Value] {
  private val log = LoggerFactory.getLogger(getClass)
  private var subscription: Flow.Subscription = _
  private var subscriber: Flow.Subscriber[_ >: Value] = _
  private val initialized = new AtomicBoolean(false)
  @volatile private var snapshotTS: Long = _

  // We do not request data from the publisher until we have one subscriber
  // to avoid discarding events before the subscriber had the chance to subscribe.
  override def subscribe(subscriber: Flow.Subscriber[_ >: Value]): Unit = {
    if (this.subscriber == null) {
      this.subscriber = subscriber
      super.subscribe(subscriber)
      requestOne()
    } else
      throw new IllegalStateException("SnapshotEventFlowProcessor can have only one subscriber")
  }

  override def onSubscribe(subscription: Flow.Subscription): Unit =
    this.subscription = subscription

  override def onNext(event: Value): Unit = {
    if (initialized.get()) {
      val eventTS = event("txn").to[Long].get
      if (eventTS > snapshotTS) submit(event) // ignore event older than doc. snapshot
      requestOne()
    } else {
      // not initialized receiving first element
      event("type") match {
        case VSuccess(StringV("start"), _) =>
          loadDocument().onComplete {
            case Failure(exception) =>
              onError(exception)
              subscription.cancel()
            case Success(documentSnapshot) =>
              snapshotTS = documentSnapshot("ts").to[Long].get
              // send start event first
              submit(event)
              // follow up with the snapshot event
              val documentEvent = ObjectV(
                "type" -> StringV("snapshot"),
                "txn" -> LongV(snapshotTS.toLong),
                "event" -> documentSnapshot)
              submit(documentEvent)
              initialized.set(true)
              // only request more when we are ready in order to avoid race condition
              requestOne()
          }
        case _ =>
          onError(new IllegalArgumentException(s"Stream did not begin with a `start` event but $event"))
          subscription.cancel()
      }
    }
  }

  override def onError(throwable: Throwable): Unit = {
    log.error("unrecoverable error encountered by subscription", throwable)
    subscriber.onError(throwable)
  }

  override def onComplete(): Unit = {
    log.debug("subscription completed")
    subscriber.onComplete()
  }

  private def requestOne(): Unit =
    subscription.request(1)
}

File: faunadb-java/src/main/java/com/faunadb/client/streaming/SnapshotEventFlowProcessor.java
package com.faunadb.client.streaming;

import com.faunadb.client.query.Expr;
import com.faunadb.client.types.Field;
import com.faunadb.client.types.Value;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Flow;
import java.util.concurrent.SubmissionPublisher;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Function;

public class SnapshotEventFlowProcessor extends SubmissionPublisher<Value> implements Flow.Processor<Value, Value> {

    public SnapshotEventFlowProcessor(Expr target, Function<Expr, CompletableFuture<Value>> loadDocument) {
        this.loadDocument = loadDocument;
        this.target = target;
    }

    private static Field<Long> TxnField = Field.at("txn").to(Long.class);
    private static Field<Long> TsField = Field.at("ts").to(Long.class);
    private static Field<String> TypeField = Field.at("type").to(String.class);

    private final Logger log = LoggerFactory.getLogger(getClass());
    private Function<Expr, CompletableFuture<Value>> loadDocument;
    private Expr target;
    private Flow.Subscription subscription = null;
    private Flow.Subscriber<? super Value> subscriber = null;
    private Long snapshotTS = null;
    private AtomicBoolean initialized = new AtomicBoolean(false);

    private void requestOne() {
        subscription.request(1);
    }

    @Override
    public void subscribe(Flow.Subscriber<? super Value> subscriber) {
        if (this.subscriber == null) {
            this.subscriber = subscriber;
            super.subscribe(subscriber);
            requestOne();
        } else {
            throw new IllegalStateException("SnapshotEventFlowProcessor can have only one subscriber");
        }
    }

    @Override
    public void onSubscribe(Flow.Subscription subscription) {
        this.subscription = subscription;
    }

    @Override
    public void onNext(Value event) {
        if (initialized.get()) {
            Long eventTS = event.get(TxnField);
            if (eventTS > snapshotTS) submit(event); // ignore event older than doc. snapshot
            requestOne();
        } else {
            // not initialized receiving first element
            boolean isStartEvent = event.getOptional(TypeField).map(s -> s.equals("start")).orElse(false);
            if (isStartEvent) {
                loadDocument.apply(target).whenComplete((documentSnapshot, ex) -> {
                    if (ex != null) {
                        onError(ex);
                        subscription.cancel();
                    } else {
                        snapshotTS = documentSnapshot.get(TsField);
                        // send start event first
                        submit(event);
                        // follow up with the snapshot event
                        Map<String, Value> fields = new HashMap<>();
                        fields.put("type", new Value.StringV("snapshot"));
                        fields.put("txn", new Value.LongV(snapshotTS));
                        fields.put("event", documentSnapshot);
                        Value documentEvent = new Value.ObjectV(fields);
                        submit(documentEvent);
                        initialized.set(true);
                        // only request more when we are ready in order to avoid race condition
                        requestOne();
                    }
                });
            } else {
                onError(new IllegalArgumentException("Stream did not begin with a `start` event but with" + event));
                subscription.cancel();
            }
        }
    }

    @Override
    public void onError(Throwable throwable) {
        log.error("unrecoverable error encountered by subscription", throwable);
        subscriber.onError(throwable);
    }

    @Override
    public void onComplete() {
        log.debug("subscription completed");
        subscriber.onComplete();
    }
}
